### PR TITLE
separate threads for admin and proxy

### LINKED ISSUE #14: Research optimal threading configuration, and implement it.
Likely a dedicated admin thread, and 1-2 worker threads.

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
Admin（debug, xds client) use one thread

proxy use by default 2 threads.

fix #14

```
root@opensrc:~/rust/ztunnel# top -H -p 844183
top - 10:23:41 up 5 days, 18:37,  3 users,  load average: 0.17, 0.27, 0.24
Threads:   5 total,   0 running,   5 sleeping,   0 stopped,   0 zombie
%Cpu(s):  1.0 us,  0.4 sy,  0.0 ni, 98.5 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem :  16012.5 total,   4429.1 free,   6529.1 used,   5054.2 buff/cache
MiB Swap:      0.0 total,      0.0 free,      0.0 used.   8996.5 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                                                                                              
 844183 root      20   0  367912  18312  14744 S   0.0   0.1   0:00.20 ztunnel                                                                                                              
 844497 root      20   0  367912  18312  14744 S   0.0   0.1   0:00.13 tokio-runtime-w                                                                                                      
 844498 root      20   0  367912  18312  14744 S   0.0   0.1   0:00.00 ztunnel                                                                                                              
 844499 root      20   0  367912  18312  14744 S   0.0   0.1   0:00.00 proxy                                                                                                                
 844500 root      20   0  367912  18312  14744 S   0.0   0.1   0:00.00 proxy   
```

### CODE REVIEW

File: src/admin/mod.rs [hunk 1]
```diff
@@ -99,11 +101,11 @@ impl Server {
         self.addr
     }
 
-    pub fn spawn(self, shutdown: &signal::Shutdown, drain_rx: Watch) {
+    pub async fn spawn(self, drain_rx: Watch) {
         let _dx = drain_rx.clone();
         let ready = self.ready.clone();
         let workload_info = self.workload_info.clone();
-        let shutdown_trigger = shutdown.trigger();
+        let shutdown_trigger = self.shutdown_trigger.clone();
```
Comment (zhlsunshine) [inline src/admin/mod.rs:108]: In fact, there would be a new shutdown_trigger created when calling `shutdown.trigger()`, so I think it's unnecessary to call the `clone()` for it.

Comment (hzxuzhonghu) [reply src/admin/mod.rs:108]: We cannot if you look at L155

Comment (zhlsunshine) [reply src/admin/mod.rs:108]: yeah, how about to use another variable? I remember that the `trigger()` will clone a trigger.

File: src/app.rs [hunk 2]
```diff
@@ -52,22 +51,33 @@ pub async fn build_with_cert(
...
             error!("workload manager: {}", e);
         }
-    }));
-    tasks.push(tokio::spawn(proxy.run()));
+    });
+
+    let proxy_addresses = proxy.addresses();
+    let worker_num = config.num_worker_threads;
+    thread::spawn(move || {
+        let runtime = tokio::runtime::Builder::new_current_thread()
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply src/app.rs:64]: This is by mistake

File: src/main.rs [hunk 3]
```diff
@@ -27,8 +27,7 @@ use ztunnel::*;
 fn main() -> anyhow::Result<()> {
     telemetry::setup_logging();
     let config: config::Config = Default::default();
-    tokio::runtime::Builder::new_multi_thread()
-        .worker_threads(config.num_worker_threads)
+    tokio::runtime::Builder::new_current_thread()
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply src/main.rs:30]: will look into it

File: src/admin/mod.rs (hunk 1 above)
Comment (hzxuzhonghu) [reply src/admin/mod.rs:152]: I think it is no difference.

**Review by howardjohn — COMMENTED** [review]
Still getting `hyper::server::tcp: accept error: IO driver has terminated`

**Comment by hzxuzhonghu** [comment]
With #171, this error does not exist

**Comment by hzxuzhonghu** [comment]
rebased

File: src/app.rs [hunk 4]
```diff
@@ -51,12 +50,13 @@ pub async fn build_with_cert(
...
 
+    let shutdown_trigger = shutdown.trigger();
     let admin = admin::Builder::new(config.clone(), workload_manager.workloads(), ready)
-        .bind(registry)
+        .bind(registry, shutdown_trigger)
         .expect("admin server starts");
     let admin_address = admin.address();
-    admin.spawn(&shutdown, drain_rx.clone());
 
+    let drain_rx_amdin = drain_rx.clone();
```
Comment (keithmattix) [inline src/app.rs:59]:
```suggestion
    let drain_rx_admin = drain_rx.clone();
```

Comment (hzxuzhonghu) [reply src/app.rs:59]: good catch

File: src/app.rs [hunk 5]
```diff
@@ -100,17 +115,12 @@ pub struct Bound {
     pub proxy_addresses: proxy::Addresses,
 
     pub shutdown: signal::Shutdown,
-    tasks: Vec<JoinHandle<()>>,
     config: config::Config,
     drain_tx: drain::Signal,
 }
 
 impl Bound {
     pub async fn spawn(self) -> anyhow::Result<()> {
```
Comment (howardjohn) [inline src/app.rs:123]: Given we already spawned the tasks and now this just waits for shutdown maybe we need to call this `await_termination` or something

**Review by howardjohn — APPROVED** [review]

File: src/app.rs (hunk 5 above)
Comment (hzxuzhonghu) [reply src/app.rs:124]: Will think about it

**Pull request merged**